src/WaveClip.cpp

```diff
@@ -1789,3 +1712,233 @@ const wxString& WaveClip::GetName() const
+{
+    return s.as_double() / mRate;
+}
+
+void WaveClip::SetSilence(sampleCount offset, sampleCount length)
+{
+    GetSequence()->SetSilence(sampleCount(floor(GetTrimLeft() * GetRate())) + offset, length);
+    MarkChanged();
+}
+
+sampleCount WaveClip::GetSequenceSamplesCount() const
+{
+    return mSequence->GetNumSamples();
+}
+
+double WaveClip::GetPlayStartTime() const noexcept
+{
+    return mSequenceOffset + mTrimLeft;
+}
+
+void WaveClip::SetPlayStartTime(double time)
+{
+    SetSequenceStartTime(time - mTrimLeft);
+}
+
+double WaveClip::GetPlayEndTime() const
+{
+    auto numSamples = mSequence->GetNumSamples();
+
+    double maxLen = GetSequenceStartTime() + ((numSamples + mAppendBufferLen).as_double()) / mRate - mTrimRight;
+    // JS: calculated value is not the length;
+    // it is a maximum value and can be negative; no clipping to 0
+
+    return maxLen;
+}
+
+sampleCount WaveClip::GetPlayStartSample() const
+{
+    return TimeToSamples(GetPlayStartTime());
+}
+
+sampleCount WaveClip::GetPlayEndSample() const
+{
+    return GetPlayStartSample() + GetPlaySamplesCount();
+}
+
+sampleCount WaveClip::GetPlaySamplesCount() const
+{
+    return mSequence->GetNumSamples()
+       - TimeToSamples(mTrimRight) - TimeToSamples(mTrimLeft);
+}
+
+void WaveClip::SetTrimLeft(double trim)
+{
+    mTrimLeft = std::max(.0, trim);
+}
+
+double WaveClip::GetTrimLeft() const noexcept
+{
+    return mTrimLeft;
+}
+
+void WaveClip::SetTrimRight(double trim)
+{
+    mTrimRight = std::max(.0, trim);
+}
+
+double WaveClip::GetTrimRight() const noexcept
+{
+    return mTrimRight;
+}
+
+void WaveClip::TrimLeft(double deltaTime)
+{
+    mTrimLeft += deltaTime;
+}
+
+void WaveClip::TrimRight(double deltaTime)
+{
+    mTrimRight += deltaTime;
+}
+
+void WaveClip::TrimLeftTo(double to)
+{
+    mTrimLeft = std::clamp(to, GetSequenceStartTime(), GetPlayEndTime()) - GetSequenceStartTime();
+}
+
+void WaveClip::TrimRightTo(double to)
+{
+    mTrimRight = GetSequenceEndTime() - std::clamp(to, GetPlayStartTime(), GetSequenceEndTime());
+}
+
+double WaveClip::GetSequenceStartTime() const noexcept
+{
+    // JS: mSequenceOffset is the minimum value and it is returned; no clipping to 0
+    return mSequenceOffset;
+}
+
+void WaveClip::SetSequenceStartTime(double startTime)
+{
+    mSequenceOffset = startTime;
+    mEnvelope->SetOffset(startTime);
+}
+
+double WaveClip::GetSequenceEndTime() const
+{
+    auto numSamples = mSequence->GetNumSamples();
+
+    double maxLen = GetSequenceStartTime() + (numSamples + mAppendBufferLen).as_double() / mRate;
+    // JS: calculated value is not the length;
+    // it is a maximum value and can be negative; no clipping to 0
+
+    return maxLen;
+}
+
+sampleCount WaveClip::GetSequenceStartSample() const
+{
+    return TimeToSamples(mSequenceOffset);
+}
+
+sampleCount WaveClip::GetSequenceEndSample() const
+{
+    return GetSequenceStartSample() + mSequence->GetNumSamples();
```

Comment:
Also not accounting for buffer length, inconsistently with `GetSequenceEndTime()`

But that was also so in removed `GetEndSample`